### PR TITLE
 [BUGFIX] Proper storage of taxes for an order

### PR DESCRIPTION
**Pull request by rintisch**
Render fields tax and orderTax in BE List module.
- Remove obsolete TCA settings canNotCollapse.
- Proper TCA configuration for taxes by introducing foreign_match_fields to correctly resolve relations.
- Add EventListener to persist taxes during an order process.
- Remove code which created unrelated tax records when ordering a product with BE variants.

Fixes: https://github.com/extcode/cart/issues/312, https://github.com/extcode/cart/issues/509

### CODE REVIEW

File: Classes/EventListener/Order/Create/PersistOrder/Products.php [hunk 1]
```diff
@@ -185,15 +184,6 @@ protected function addVariantsOfVariant(BeVariant $variant, int $level): void
...
     protected function addBeVariant(BeVariant $variant, int $level): void
     {
-        $orderTax = GeneralUtility::makeInstance(
-            Tax::class,
-            $variant->getTax(),
-            $this->taxClasses[$variant->getTaxClass()->getId()]
-        );
-        $orderTax->setPid($this->storagePid);
-
-        $this->taxRepository->add($orderTax);
```
Comment (extcode) [inline Classes/EventListener/Order/Create/PersistOrder/Products.php:195]: Think the `$this->taxRepository` isn't needed anymore in this class. Could the property and inject-method can also be removed?

File: Classes/EventListener/Order/Create/PersistOrder/Taxes.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        $orderItem = $this->addTaxToRepositoryAndItem($event, $orderItem, 'tax');
+        $orderItem = $this->addTaxToRepositoryAndItem($event, $orderItem, 'totalTax');
+
+        $this->itemRepository->update($orderItem);
+        $this->persistenceManager->persistAll();
+    }
+
+    protected function addTaxToRepositoryAndItem(
+        PersistOrderEventInterface $event,
+        OrderItem $orderItem,
```
Comment (extcode) [inline Classes/EventListener/Order/Create/PersistOrder/Taxes.php:43]: Can we remove the extra function parameter $orderItem? We can get it from `$event` in this method again.
```suggestion
        
```

File: Classes/EventListener/Order/Create/PersistOrder/Taxes.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        string $typeOfTax
+    ): OrderItem {
+        $cart = $event->getCart();
```
Comment (extcode) [inline Classes/EventListener/Order/Create/PersistOrder/Taxes.php:46]: See comment above.

```suggestion
        $cart = $event->getCart();
        $orderItem = $event->getOrderItem();
```

File: Classes/EventListener/Order/Create/PersistOrder/Taxes.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        $storagePid = $event->getStoragePid();
+        $taxClasses = $event->getTaxClasses();
+
+        $taxes = $typeOfTax === 'tax' ? $cart->getTaxes() : $cart->getTotalTaxes();
+
+        /** @var int $taxClassId */
+        /** @var float $tax */
+        foreach ($taxes as $taxClassId => $tax) {
```
Comment (extcode) [inline Classes/EventListener/Order/Create/PersistOrder/Taxes.php:54]: If you change:
```suggestion
        foreach ($taxes as $taxClassId => $taxValue) {
```
you can remove the new variable in line 55.

File: Classes/EventListener/Order/Create/PersistOrder/Taxes.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            $taxValue = $tax;
+            $taxClass = $taxClasses[$taxClassId];
+            $orderTax = GeneralUtility::makeInstance(OrderTax::class, $taxValue, $taxClass);
```
Comment (extcode) [inline Classes/EventListener/Order/Create/PersistOrder/Taxes.php:57]: I think this is also readable:
```suggestion
            $orderTax = GeneralUtility::makeInstance(OrderTax::class, $taxValue, $taxClasses[$taxClassId]);
```
to we do not need the variable assignment.

File: Configuration/Services.yaml [hunk 6]
```diff
@@ -49,6 +49,13 @@ services:
         event: Extcode\Cart\Event\Order\PersistOrderEvent
         after: 'cart--order--create--persist-order--item'
 
+  Extcode\Cart\EventListener\Order\Create\PersistOrder\Taxes:
+    tags:
+      - name: event.listener
+        identifier: 'cart--order--create--persist-order--tax'
```
Comment (extcode) [inline Configuration/Services.yaml:55]: My syntax for the identifier is related to the class name. Can you change this?

```suggestion
        identifier: 'cart--order--create--persist-order--taxes'
```

File: Classes/EventListener/Order/Create/PersistOrder/Products.php (hunk 1 above)
Comment (rintisch) [reply Classes/EventListener/Order/Create/PersistOrder/Products.php:195]: @extcode I have unfortunately no laptop with me, just a smartphone with the GitHub app. If you want to merge it quick you need to adapt the code yourself. All the proposals look good to me.

**Pull request merged**